### PR TITLE
add user engagement metric

### LINKED ISSUE #240: Create API endpoint for "User engagement" metrics
See #215 

- How many users saved at least one new resource within a given time period
- How many users saved at least one new template within a given time period

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made?

Fixes #240 - adding user engagement metric query

## How was this change tested?

Added new test

### CODE REVIEW

File: src/endpoints/metrics.js [hunk 1]
```diff
@@ -26,6 +26,19 @@ const getResourceQuery = (resourceType) => {
...
+ * @param {string} endDate The end date
+ * @returns {object} The query to send to mongo to filter by the specified dates
+ */
+const getDateQuery = (startDate, endDate) => {
+  return {
+    $gt: new Date(startDate),
+    $lt: new Date(endDate),
+  }
+}
+
```
Comment (peetucket) [inline src/endpoints/metrics.js:41]: A small refactor to make this query syntax more DRY.  Need to port over to #242 as well (both are in review right now, so may depend on the sequencing of how they get merged)

File: src/endpoints/metrics.js [hunk 2]
```diff
@@ -58,10 +81,10 @@ metricsRouter.get("/createdCount/:resourceType", (req, res, next) => {
     { $unwind: "$resourceMetadata" },
     {
       $match: {
-        "resourceMetadata.versions.0.timestamp": {
-          $gt: new Date(req.query.startDate),
-          $lt: new Date(req.query.endDate),
-        },
+        "resourceMetadata.versions.0.timestamp": getDateQuery(
```
Comment (mjgiarlo) [inline src/endpoints/metrics.js:84]: Not exactly relevant to this PR, but can you say more about the `resourceMetadata.versions.0` part of the query?

File: src/endpoints/metrics.js (hunk 1 above)
Comment (peetucket) [inline src/endpoints/metrics.js:41]: Small refactor to DRY this up, since we now use this filter in three different endpoints

**Review by mjgiarlo — APPROVED** [review]
LGTM. I have one question that is not really about what's in this PR, and it's non-blocking. Merge at will.

File: src/endpoints/metrics.js (hunk 2 above)
Comment (peetucket) [reply src/endpoints/metrics.js:84]: yes, will add a comment

**Comment by peetucket** [comment]
Note: I'm pulling this back to draft because I want to verify that the response from the new `distinct` query is really the same as the response from the `aggregate` queries, it may not be.

**Comment by peetucket** [comment]
There was a problem with the query, and the parsing of the response, both of which I fixed (and updated the test to address).

File: src/endpoints/metrics.js [hunk 3]
```diff
@@ -6,6 +6,12 @@ const metricsRouter = express.Router()
 // Add the db to req
 metricsRouter.use(connect)
 
+// NOTE:
+// When filtering for created resources by time, we want to only look at the very first timestamp in the versions element in
+//  `resourceMetadata`. This is what `0` does in `resourceMetadata.versions.0.timestamp`
+//  (only filter on the first timestamp in that array).
+//  When filtering for edited resources, we want to look at all timestamps in that array.
+
```
Comment (peetucket) [inline src/endpoints/metrics.js:14]: Comment added to code to explain that odd `0` syntax.

**Review by justinlittman — APPROVED** [review]

**Pull request merged**